store/storage/db.go

```diff
@@ -0,0 +1,221 @@
+package storage
+
+import (
+	"fmt"
+	"sync"
+
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+
+	"cosmossdk.io/store/v2"
+)
+
+// TODO:
+//
+// - Define and implement pruning method(s)
+// - Import/snapshot method(s)???
+// - Should we make this an interface?
+//
+// Ref: https://github.com/cosmos/cosmos-sdk/issues/17279
+
+// DefaultMemDBSize defines the default pre-allocation size of the in-memory
+// database used to accumulate database entries prior to committing them to disk.
+const DefaultMemDBSize = 1000
+
+var (
+	_ store.VersionedReaderWriter    = (*Database)(nil)
+	_ store.VersionedIteratorCreator = (*Database)(nil)
+	_ store.Committer                = (*Database)(nil)
+)
+
+// Database defines the state storage (SS) backend. Each open instance of a Database
+// is meant to commit key/value pairs to disk upon Commit for a single version.
+// However, it can be used to read key/value pairs at any version.
+type Database struct {
+	mu      sync.RWMutex
+	vdb     store.VersionedDatabase
+	version uint64
+	memSize int
+	mem     map[string]map[string]dbEntry
```

Comment:
could you explain the use-case you see for this cache?  SS will be filled by the store package async, I can't see how outside consumers would need to write or need to read from staged writes.

Reply:
Writes in SS are buffered until `Commit` is called. What's missing from the API (which I'll implement later) is a `Branch` or `Copy` API, which is analogous to the caching/branching that you see today.

> SS will be filled by the store package async, I can't see how outside consumers would need to write or need to read from staged writes.

I'm not sure I understand this statement? The root multi-store will have access to a single SS `*Database`. This acts the QMS that will exist in `BaseApp`. Outside consumers, e.g. `BaseApp` or modules work directly with this type.

Reply:
>Outside consumers, e.g. BaseApp or modules work directly with this type, reading and writing to/from it via the root multi-store.

In what scenario does a module or BaseApp need to write a version n to change history?  Module updates go into the latest version, SC, not SS.

Reply:
I think some design around data flow and timing is missing.  Are you envisioning that writes to SS be synchronous with SC? That will (probably) drastically slow down block time for nodes with bigger storage.  I had envisioned _n_ recent block reads/writes routing to SC while SS is filled in a background thread. 
If that's the case there's no need for a queryable cache, and the fill thread would just handle batching, no consumer API needed.